design/dedicated-osd-pod.md

```diff
@@ -0,0 +1,66 @@
+# Run OSD in its own Pod
+
+## TL;DR
+
+A one-OSD-per-Pod placement should be implemented to improve reliability and resource efficiency for Ceph OSD daemon.
+
+## Background
+
+Currently in Rook 0.7, Rook Operator starts a ReplicaSet to run [`rook osd`](https://github.com/rook/rook/blob/master/cmd/rook/osd.go) command (hereafter referred to as `OSD Provisioner`)  on each storage node. The ReplicaSet has just one replica. `OSD Provisioner` scans and prepares devices, creates OSD IDs and data directories or devices, generates Ceph configuration. At last, `OSD Provisioner` starts all Ceph OSD, i.e. `ceph-osd`, daemons in foreground and tracks `ceph-osd` processes.
+
+As observed, all Ceph OSDs are running in the same Pod.
+
+## Limitations
+
+The limitations of current design are:
+
+- Reliability issue. One Pod for all OSDs doesn't have the highest reliability nor efficiency. If the Pod is deleted, accidentally or during maintenence, all OSDs are down till the ReplicaSet restart. 
+- Efficiency issue. Resource limits cannot be set effectively on the OSDs since the number of osds per in the pod could vary from node to node. The operator cannot make decisions about the topology because it doesn't know in advance what devices are available on the nodes.
+- Process management issue. Rook's process management is very simple. Using Kubernetes pod management is much more reliable.
+
+
+A more comprehensive discussions can be found at [this issue](https://github.com/rook/rook/issues/1341).
+
+## Proposal   
+
+We propose the following change to address the limitations.
+
+### Terms
+
+- OSD Discovery. A process that is given a device or directory path filter and use the filter to discover devices on the host. The process updates the OSD orchestration Configmap with the device paths and mark the orchestration status as "DeviceDiscovered". The process is created by Rook Operator and runs in a Kubernetes deployment on the storage node. This process is storage backend agnostic. Currently `rook osd` command implements this functionality.
```

Comment:
i'm wondering if this device discovery should not take into account any filters at all.  Should it instead just discover everything on the node and report it in the config map?  What's the advantage of filtering out during the discovery phase, instead of discovering everything and then performing filtering within the operator when it's determining what OSDs to deploy?  It sounds like the operator will need to perform filtering anyways if the filters change, so perhaps the filtering should only be done in the operator, not within the discovery.

Reply:
agreed, filtering should be done by the operator and cluster crd after discovery

Reply:
The advantage of filtering at the device discovery rather than at operator is that, it is much easy and extensible to build a filter that runs on the storage nodes by device discovery pod than having such capability in operator that doesn't have access to the storage devices.

Consider the following:
- how to discover *everything* on the nodes. Inevitably, there must be some hard coded logic there to walk the device tree and find anything appropriate based on some assumptions. If the assumptions deviate from the host configuration, then not every device can be discovered. This is especially true when discovering devicemapper targets, that have variable names and on hosts that have different udev rules.

- even if discovering everything is possible, at what level of details should device discovery pod tell operator about the device? For instance, if the filter is to discover all devices made from Vendor X with Model Y, the device discovery pod can run the filtering script on the host to find them without code change. This flexibility cannot be achieved by operator if device discovery pod fails to pass every information about the device to the operator.

Admittedly there are use cases that users may change the filter on the fly and it may be too late for device discovery pod to pick up the new filters. But this can be done by storing filters in configmap and device discovery pod loads filter from configmap each time it polls the device. Then devices under new filters will eventually be discovered.

Reply:
@rootfs how about if we keep the device discovery simple and just discover what is under `/dev`? In the near future, we will depend on the local pvs, which would hopefully give more flexibility in what we can discover and consume. 
I was thinking of the filtering as the decision point about which devices should be used (specified by the cluster crd), rather than using filtering to increase our capabilities during discovery.